### PR TITLE
Add support for stability analysis

### PR DESCRIPTION
**Pull request by atomb**
Add two new options for stability analysis

The `--max-resource-stddev` and `--max-duration-stddev` flags to the `summarize-csv-results` command set the maximum value of the standard deviation of the resource count and duration measurements, respectively.

Since each verification execution generates a new CSV file, and the report generator reads all CSV files found, by default, we can just run verification several times to have several measurements for each Dafny proof obligation batch.

### CODE REVIEW

File: src/TestResult.dfy
Comment (atomb) [inline src/TestResult.dfy:68]: It probably makes more sense to have `f` be of type `TestResult -> real`, and convert from ticks to seconds _before_ calculating statistics for duration measurements. But `Min`, `Max`, and `Map` are specific to `seq<int>`, so we'd have to introduce custom versions of them.

File: .github/workflows/build-and-test.yml
Comment (cpitclaudel) [inline .github/workflows/build-and-test.yml:41]: rename to `seed`?

File: .github/workflows/build-and-test.yml
Comment (cpitclaudel) [inline .github/workflows/build-and-test.yml:42]: Why not `dotnet run` with the Dafny directory?

File: src/Externs.cs
Comment (cpitclaudel) [inline src/Externs.cs:70]: Are `num` and `den` private?  If they are less use reflection to get to them for now (https://stackoverflow.com/questions/3303126/how-to-get-the-value-of-private-field-using-reflection/3303182#3303182); then let's patch Dafny.  But we should not be parsing strings, I think

File: src/Main.dfy
Comment (cpitclaudel) [inline src/Main.dfy:135]: Can we merge these two functions?

File: src/Statistics.dfy
Comment (cpitclaudel) [inline src/Statistics.dfy:13]: We really should have a version of sum without the starting parameter.  Not in this PR though.

File: src/Statistics.dfy
Comment (cpitclaudel) [inline src/Statistics.dfy:25]: Technically xbar?

File: src/Statistics.dfy
Comment (cpitclaudel) [inline src/Statistics.dfy:26]: Does this need a `-1` in the denominator?

File: src/Statistics.dfy
Comment (cpitclaudel) [inline src/Statistics.dfy:47]: Need to update the stddev as well?

File: src/TestResult.dfy
Comment (cpitclaudel) [inline src/TestResult.dfy:66]: Bonus points if you prove this equivalent to this:

```
method groupBy'<T(==)>(results: seq<T>, f: T -> string) returns (m: map<string, set<T>>) {
  var keys := set t: T | t in results :: f(t);
  m := map k: string | k in keys :: (set t: T | t in results && k == f(t));
}
```

File: src/TestResult.dfy
Comment (cpitclaudel) [reply src/TestResult.dfy:68]: Why not change Min, Max, and Map to work with reals and do everything in terms for reals?

Comment (cpitclaudel) [reply src/TestResult.dfy:68]: Or are they library functions?  In that case we need MaxF that takes a comparison function

File: src/TestResult.dfy
Comment (cpitclaudel) [inline src/TestResult.dfy:115]: A ma p would be this:

```
function method MapMap<K, V, V'>(m: map<K, V>, f: V -> V') : map<K, V'> {
  map k: K | k in m :: f(m[k])
}
```

But you want a fold, right?  I don't know of a good way to do that :'(

**Review by cpitclaudel — DISMISSED** [review]
LGTM, left some comments that could just be FIXMEs for the future

File: .github/workflows/build-and-test.yml
Comment (atomb) [reply .github/workflows/build-and-test.yml:41]: Good idea.

File: .github/workflows/build-and-test.yml
Comment (atomb) [reply .github/workflows/build-and-test.yml:42]: This is slightly simpler?

File: src/Externs.cs
Comment (atomb) [reply src/Externs.cs:70]: Yeah, they are. We should update the C# runtime to make them accessible, but that'll need to wait for a Dafny release, so this is a stopgap. I hope it won't stick around for more than a few weeks.

File: src/Statistics.dfy
Comment (atomb) [reply src/Statistics.dfy:13]: Agreed.

File: src/Statistics.dfy
Comment (atomb) [reply src/Statistics.dfy:47]: I did that in an earlier PR, but it's not as simple as dividing. I think we really should convert to seconds before doing the statistical calculations (see your other comments).

File: src/TestResult.dfy
Comment (atomb) [reply src/TestResult.dfy:66]: Maybe a nice thing to do soon! But in a future PR.

File: src/TestResult.dfy
Comment (atomb) [reply src/TestResult.dfy:115]: Yeah, they're library functions. We really need traits on primitive and data types so that we can do arithmetic over sequences of arbitrary numeric things. :)

And, yes, that really is a fold! I'll update the comment.

File: src/Main.dfy
Comment (atomb) [reply src/Main.dfy:135]: Yeah, we really could. I'll do that.

File: .github/workflows/build-and-test.yml
Comment (cpitclaudel) [reply .github/workflows/build-and-test.yml]: OK.  Bu tare we sure it picks up the right Dafny?  I guess there's only one anyway.

File: src/Externs.cs
Comment (cpitclaudel) [reply src/Externs.cs:70]: Why not just reflection in the meantime?

File: src/Statistics.dfy
Comment (cpitclaudel) [reply src/Statistics.dfy:47]: I think it actually is? `stddev(a*X)` is `abs(a) * stddev(X)`

File: src/Externs.cs
Comment (atomb) [reply src/Externs.cs:70]: It seems similarly awkward to me? I guess it's probably faster. Given that this is in place and works, I'd be more inclined to focus on making neither one necessary. :)

File: .github/workflows/build-and-test.yml
Comment (atomb) [reply .github/workflows/build-and-test.yml]: Yeah, there should only be one.

File: src/Statistics.dfy
Comment (atomb) [reply src/Statistics.dfy:47]: Ah, right. I hadn't done the algebra and didn't think it'd work out that cleanly!

File: src/Externs.cs
Comment (cpitclaudel) [reply src/Externs.cs:70]: It's simple and reliable (not pretty, but we know if works).  This string parsing scares me.  

Not a big deal, let's make a PR on he heels of this one to expose AsDouble.

**Review by cpitclaudel — APPROVED** [review]

**Review by MikaelMayer — COMMENTED** [review]
Good job.
A few suggestions that I wasn't fast enough to push before you merged them.

File: src/TestResult.dfy
Comment (atomb) [reply src/TestResult.dfy:70]: Oh, right! I had meant to change that. :)

File: src/TestResult.dfy
Comment (atomb) [reply src/TestResult.dfy:109]: It would be nice to prove, but I'm not yet sure if it's worth the effort. We'd need to track that we have a map full of sequences that are all non-empty.

File: src/TestResult.dfy
Comment (atomb) [reply src/TestResult.dfy]: Yeah, I think this is a nice way to go, and gives us some good things to add to the standard library.

Comment (atomb) [reply src/TestResult.dfy]: I'll probably do more stuff before it's ready to review, but I've started in this direction: https://github.com/dafny-lang/dafny-reportgenerator/commit/563a56b4368a4bdf47231c501ce68c70e39295aa

File: src/TestResult.dfy
Comment (MikaelMayer) [reply src/TestResult.dfy:109]: What about this:
```
type NonEmptySeq<T(0)> = x : seq<T> | |x| > 0 witness *
...

method PrintAllTestResultStatistics(groupedResults: map<string, NonEmptySeq<TestResult>>)

....
method GroupTestResults(results: seq<TestResult>) returns (groupedResults: map<string, NonEmptySeq<TestResult>>) {
```
No need to keep track of invariants, subsets type can do it for you.
I wish we could write an explicit witness but I don't know how.

**Pull request merged**
